AX: Eagerly computing AXProperty::Orientation is extremely expensive and unnecessary

twilco opened the pull request:
#### 412bf3c1d45cb9473eff5561c62a9d0baa98112b
<pre>
AX: Eagerly computing AXProperty::Orientation is extremely expensive and unnecessary
<a href="https://bugs.webkit.org/show_bug.cgi?id=290228">https://bugs.webkit.org/show_bug.cgi?id=290228</a>
<a href="https://rdar.apple.com/147621116">rdar://147621116</a>

Reviewed by Chris Fleizach.

Eagerly computing AXProperty::Orientation adds 12 seconds of time to the initial build of the accessibility tree
on <a href="http://html.spec.whatwg.org">http://html.spec.whatwg.org</a> because for most objects, it requires computing elementRect(), which is very expensive.

This commit resolves the issue by separating &quot;explicit&quot; orientations (e.g. those from aria-orientation, scrollbars, etc)
into a new explicitOrientation() function, leaving orientation() as a fallback to handle any &quot;implicit&quot; orientations
(e.g. those that default from other state, like role value or size).

This greatly reduces the work to cache this property, and saves memory, since we used to store AXProperty::Orientation
for every object, and now only do so in the rare cases we have an explicitly defined orientation.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::orientation const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::orientationFromARIA const):
(WebCore::AccessibilityNodeObject::orientation const): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::orientation const): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityScrollbar.cpp:
(WebCore::AccessibilityScrollbar::explicitOrientation const):
(WebCore::AccessibilityScrollbar::orientation const): Deleted.
* Source/WebCore/accessibility/AccessibilityScrollbar.h:
* Source/WebCore/accessibility/AccessibilitySlider.cpp:
(WebCore::AccessibilitySlider::explicitOrientation const):
(WebCore::AccessibilitySlider::orientation const): Deleted.
* Source/WebCore/accessibility/AccessibilitySlider.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::setProperty):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):

Canonical link: <a href="https://commits.webkit.org/292616@main">https://commits.webkit.org/292616@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1ba1648c1ac564565377345c6d0f0ca0e529ff5b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96309 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15923 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5895 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101376 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46828 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98354 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16219 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24356 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73405 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30635 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99312 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12166 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86997 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53742 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11922 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4764 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46155 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4862 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103404 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23376 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82445 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23627 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83022 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81822 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20598 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26459 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3893 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16750 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23339 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28494 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22998 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26478 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24739 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -761,6 +761,33 @@ bool AXCoreObject::selfOrAncestorLinkHasPopup() const
+        return AccessibilityOrientation::Horizontal;
+
+    // Lacking concrete evidence of orientation, horizontal means width > height. vertical is height > width;
+    auto size = this->size();
```
Should we use `relativeFrame()` over size? The increase in precision with it being a `FloatRect` might disambiguate a few more cases (none should be super critical, however).

twilco replied:
Size is implemented with `relativeFrame`:

`IntSize size() const final { return snappedIntRect(LayoutRect(relativeFrame())).size(); }`

twilco replied:
Oh I see what you're saying. I think that would technically be slightly more accurate, but I think in this case it may be best to not deviate from the current behavior too much (`AccessibilityObject`s would start calling `relativeFrame` instead of `elementRect`, and the former does unnecessary coordinate-conversion work which is not needed for getting the size)

hoffmanjoshua replied:
Makes sense —definitely not necessary in that case

twilco committed:
AX: Eagerly computing AXProperty::Orientation is extremely expensive and unnecessary
https://bugs.webkit.org/show_bug.cgi?id=290228
rdar://147621116

Reviewed by Chris Fleizach.

Eagerly computing AXProperty::Orientation adds 12 seconds of time to the initial build of the accessibility tree
on http://html.spec.whatwg.org because for most objects, it requires computing elementRect(), which is very expensive.

This commit resolves the issue by separating "explicit" orientations (e.g. those from aria-orientation, scrollbars, etc)
into a new explicitOrientation() function, leaving orientation() as a fallback to handle any "implicit" orientations
(e.g. those that default from other state, like role value or size).

This greatly reduces the work to cache this property, and saves memory, since we used to store AXProperty::Orientation
for every object, and now only do so in the rare cases we have an explicitly defined orientation.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::orientation const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::orientationFromARIA const):
(WebCore::AccessibilityNodeObject::orientation const): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::orientation const): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityScrollbar.cpp:
(WebCore::AccessibilityScrollbar::explicitOrientation const):
(WebCore::AccessibilityScrollbar::orientation const): Deleted.
* Source/WebCore/accessibility/AccessibilityScrollbar.h:
* Source/WebCore/accessibility/AccessibilitySlider.cpp:
(WebCore::AccessibilitySlider::explicitOrientation const):
(WebCore::AccessibilitySlider::orientation const): Deleted.
* Source/WebCore/accessibility/AccessibilitySlider.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::setProperty):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):

Canonical link: https://commits.webkit.org/292616@main
